cmd/hope/vm/image.go

```diff
@@ -0,0 +1,180 @@
+package vm
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+)
+
+import (
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+import (
+	"github.com/Eagerod/hope/cmd/hope/utils"
+	"github.com/Eagerod/hope/pkg/packer"
+	"github.com/Eagerod/hope/pkg/scp"
+	"github.com/Eagerod/hope/pkg/ssh"
+)
+
+var imageCmdParameterSlice *[]string
+
+func initImageCmdFlags() {
+	imageCmdParameterSlice = imageCmd.Flags().StringArrayP("param", "p", []string{}, "parameters to forward to packer's -var")
+}
+
+type PackerBuilder struct {
+	VMName          string `json:"vm_name"`
+	OutputDirectory string `json:"output_directory"`
+}
+
+type PackerSpec struct {
+	Builders []PackerBuilder
+}
+
+var imageCmd = &cobra.Command{
+	Use:   "image",
+	Short: "Creates a VM image from the defined packer spec.",
+	Args:  cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		vmName := args[0]
+
+		vms, err := utils.GetVMs()
+		if err != nil {
+			return err
+		}
+
+		vm, err := utils.VMSpec(vmName)
+		if err != nil {
+			return err
+		}
+
+		vmDir := path.Join(vms.Root, vm.Name)
+		log.Trace(fmt.Sprintf("Looking for VM definition in %s", vmDir))
+
+		stat, err := os.Stat(vmDir)
+		if err != nil && os.IsNotExist(err) {
+			return fmt.Errorf("VM spec directory (%s) not found", vmDir)
+		} else if err != nil {
+			return err
+		}
+
+		if !stat.IsDir() {
+			return fmt.Errorf("VM spec directory (%s) is just a file", vmDir)
+		}
+
+		packerJsonPath := path.Join(vmDir, "packer.json")
+		if _, err := os.Stat(packerJsonPath); err != nil && os.IsNotExist(err) {
+			return fmt.Errorf("VM packer file not found at path: %s", packerJsonPath)
+		} else if err != nil {
+			return err
+		}
+
+		log.Debug(fmt.Sprintf("Copying contents of %s for parameter replacement.", vmDir))
+		tempDir, err := utils.ReplaceParametersInDirectoryCopy(vmDir, vm.Parameters)
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(tempDir)
+
+		// Check caches to see if I even want to build this again.
+		tempPackerJsonPath := path.Join(tempDir, "packer.json")
+		bytes, err := ioutil.ReadFile(tempPackerJsonPath)
+		if err != nil {
+			return err
+		}
+
+		var packerSpec PackerSpec
+		if err := json.Unmarshal(bytes, &packerSpec); err != nil {
+			return err
+		}
+
+		// Packer runs out of temp dir, so output directory has to be
+		//   absolute.
+		packerOutDir := packerSpec.Builders[0].OutputDirectory
+		if !path.IsAbs(packerOutDir) {
+			return fmt.Errorf("Directory %s must be absolute;", packerOutDir)
+		}
+
+		if stat, err := os.Stat(packerOutDir); err == nil {
+			if stat.IsDir() {
+				files, err := ioutil.ReadDir(packerOutDir)
+				if err != nil {
+					return err
+				}
+
+				if len(files) != 0 {
+					filenames := []string{}
```

Comment:
Don't think this is super necessary. At least it prints out the path in the error message. Could just look there manually if it mattered + wasn't just an error.